salt/states/file.py

```diff
@@ -312,18 +313,36 @@ def _gen_keep_files(name, require):
     Generate the list of files that need to be kept when a dir based function
     like directory or recurse has a clean.
     '''
+    def _is_subdir(path, directory):
+        path = os.path.realpath(path)
+        directory = os.path.realpath(directory)
+
+        relative = os.path.relpath(path, directory)
+
+        return (not relative.startswith(os.pardir))
```

Comment:
yes it's a bad name, it's to check a path is under a dir or not, not to check if that is a dir.

````
import os
def is_subdir(path, directory):
    path = os.path.realpath(path)
    directory = os.path.realpath(directory)

    relative = os.path.relpath(path, directory)

    return (not relative.startswith(os.pardir))

assert(is_subdir('/etc/salt', '/etc') == True)
assert(is_subdir('/usr/bin', '/etc') == False)
````

Above ran successfully.